File: Kmeans.py
def dist(point1, point2):
    sum = 0
    d = len(point1)
    for i in range(d):
        sum += (point1[i] - point2[i]) ** 2
    sum = sum ** 0.5
    return sum


def mindist(centroids_list, point):
    min = "inf"
    index = 0
    for j in range(len(centroids_list)):
        distance = dist(centroids_list[j], point)
        if (min == "inf" or distance < min):
            min = distance
            index = j
    return index


def clustering(centroids_list, DataPoints, indexingList):
    for i in range(len(DataPoints)):
        x = mindist(centroids_list, DataPoints[i])
        indexingList[x].append(i)


def update(centroids_list, DataPoints, indexingList):
    boolean = True
    for i in range(len(centroids_list)):
        old_avg = centroids_list[i]
        centroids_list[i] = Avg(indexingList[i], DataPoints)
        distance = dist(centroids_list[i], old_avg)
        if (distance > 0.001):
            boolean = False

    return boolean


def Avg(lst, DataPoints):
    avg_point = []
    d = len(DataPoints[0])
    for i in range(d):
        sum = 0
        for item in lst:
            sum += DataPoints[item][i]
        sum = sum / len(lst)
        avg_point.append(sum)
    return avg_point


def kmeans(k, max_iter, inputfile, outputfile):

    centroids_list = []
    DataPoints = []

    file = open(inputfile)
    content = file.read()
    DataPoints = content.splitlines()
    for i in range(len(DataPoints)):
        DataPoints[i] = DataPoints[i].split(",")
        for j in range(len(DataPoints[i])):
            DataPoints[i][j] = float(DataPoints[i][j])

    for i in range(k):
        centroids_list.append(DataPoints[i])
    if(k<=1 or k>= len(DataPoints)):
        print("Invalid Input!")
        sys.exit()

    if(max_iter<1):
        print("Invalid Input!")
        sys.exit()


    for iter in range(max_iter):
        indexingList = []
        for i in range(k):
            indexingList.append([])
        clustering(centroids_list, DataPoints, indexingList)
        bol = update(centroids_list, DataPoints, indexingList)
        if (bol == True):
            break

    for i in range(len(centroids_list)):
        for j in range(len(centroids_list[0])):
            a = centroids_list[i][j]
            centroids_list[i][j] = "%.4f" %round(a, 4)



    ff = open(outputfile, 'w')
    for i in range(len(centroids_list)):
        ff.write(','.join([format(centroids_list[i][j]) for j in range(len(centroids_list[i]))]) + '\n')
    ff.close()

import sys
args = sys.argv
maxiter=0

if len(args)==5:
    if (args[1].isdigit() == False or args[2].isdigit() == False):
        print("Invalid Input!")
        sys.exit()

    assert args[1].isdigit() , "Invalid input."
    assert args[2].isdigit(), "Invalid input."
    k = int(args[1])
    maxiter=int(args[2])
    inputfile =args[3]
    outputfile =args[4]
    kmeans(k, maxiter,inputfile,outputfile)
if len(args)==4:
    if(args[1].isdigit()==False):
        print("Invalid Input!")
        sys.exit()

    k = int(args[1])
    maxiter=200
    inputfile = args[2]
    outputfile = args[3]
    kmeans(k, maxiter, inputfile, outputfile)

File: Kmeans.c
#include <stdio.h>
#include <malloc.h>
#include <math.h>
#include <stdlib.h>

int valid(int argc, char *argv[]){
    char** ptr;
    ptr=NULL;
    if(argc<4) return 0;
    else if(argc==4) {
        if (strtod(argv[1],ptr) == 0) {
            return 0;
        }
    }
    else if(argc==5){
        if(strtod(argv[1],ptr)==0 || strtod(argv[2],ptr)==0) return 0;
    }
    free(ptr);
    return 1;
}


void Sizefile(char *filename, int *dimension, int *rows) {
    int tmp_dimension;
    int tmp_rows;
    FILE *f;
    char c;
    f = fopen(filename, "rt");
    tmp_dimension=0;
    tmp_rows=0;


    while ((c = fgetc(f)) != EOF) {
        if (c == ',') {
            tmp_dimension++;
        }
        if (c == '\n') {
            tmp_rows++;
        }
    }

    tmp_dimension = tmp_dimension / tmp_rows + 1;
    fclose(f);
    *dimension = tmp_dimension;
    *rows = tmp_rows;
}

double dist(double *point1, double *point2, int dimension) {

    int i;
    double sum = 0;
    for (i = 0; i < dimension; i++) {
        sum += pow((point1[i]-point2[i]), 2);
    }

    sum = pow(sum, 0.5);
    return sum;
}

int mindist(int k, int dimension, double **centroids_list, double *point) {

    double min = dist(centroids_list[0], point, dimension);
    int index = 0;
    int j;
    for (j = 1; j < k; j++) {
        double distance = dist(centroids_list[j], point, dimension);
        if (distance < min) {
            min = distance;
            index = j;
        }
    }
    return index;
}

void Init(int k, int dimension, int *count_array, double **sum_array) {
    int i, j;
    for (i = 0; i < k; i++) {
        count_array[i] = 0;
        for (j = 0; j < dimension; j++) {
            sum_array[i][j] = 0;
        }
    }
}

void clustering(int k, int rows, int dimension, int *count_array,
                double **sum_array, double **DataPoints, double **centroids_list) {
    int i;
    int j;
    int index;
    for (i = 0; i < rows; i++) {
        index = mindist(k, dimension, centroids_list, DataPoints[i]);
        count_array[index]++;
        for (j = 0; j < dimension; j++) {
            sum_array[index][j] += DataPoints[i][j];
        }
    }

    for (i = 0; i < k; i++) {
        for (j = 0; j < dimension; j++) {
            sum_array[i][j] = sum_array[i][j] / count_array[i];
        }
    }
}

int calculate_delta(int k, int dimension, double **centroids_list, double **sum_array) {
    double distance;
    int i, j;
    int bol = 1;
    for (i = 0; i < k; i++) {
        distance = dist(centroids_list[i], sum_array[i], dimension);
        if (distance > 0.001) {
            bol = 0;
            break;
        }
    }

    for (i = 0; i < k; i++) {
        for (j = 0; j < dimension; j++) {
            centroids_list[i][j] = sum_array[i][j];
        }
    }
    return bol;
}


int kmeans(int k, int max_iter, char filename[], char *output_filename) {
    int dimension = 0;
    int rows = 0;
    int negative;
    int p = 1;
    int i, j;
    double num, fr;
    double **centroids_list;
    double **sum_array;
    double **DataPoints;
    int *count_array;
    FILE *output;
    int delta = 0;
    int iter;
    double tmp;
    FILE *file;
    char C;

    Sizefile(filename, &dimension, &rows);
    DataPoints = malloc(sizeof(double *) * rows);
    for (i = 0; i < rows; i++) {
        DataPoints[i] = (double *) malloc((dimension) * sizeof(double));
    }


    file = fopen(filename, "rt");
    for (i = 0; i < rows; i++) {
        for (j = 0; j < dimension; j++) {
            tmp = 0;
            negative = 0;
            while ((C = fgetc(file)) != '.' && C != EOF) {
                num = 1;
                if (C == '-') {
                    negative = 1;
                } else {
                    tmp = tmp*10;
                    num = num * ((double) C - 48);
                    tmp += num;
                    num = tmp;
                }

            }
            fr = 0;
            p = 1;
            while ((C = fgetc(file)) != EOF && C != ',' && C != '\n') {
                fr = fr + ((double) C - 48) * pow(10, -1 * p);
                p++;
            }
            num = num + fr;
            if (negative == 1) {
                num = -1 * num;
            }
            DataPoints[i][j] = num;
        }
    }
    fclose(file);
    if(k>rows || k<1 || max_iter<1){
        printf("Invalid Input!");
        return 1;
    }

    centroids_list = malloc(sizeof(double *) * k);
    for (i = 0; i < k; i++) {
        centroids_list[i] = (double *) malloc((dimension) * sizeof(double));
    }

    for (i = 0; i < k; i++) {
        for (j = 0; j < dimension; j++) {
            centroids_list[i][j] = DataPoints[i][j];
        }
    }

    sum_array = malloc(sizeof(double *) * k);
    for (i = 0; i<k; i++) {
        sum_array[i] = (double *) malloc((dimension) * sizeof(double));
    }

    count_array = (int *) malloc(sizeof(int) * k);

    iter = 0;
    while(delta==0 && iter < max_iter) {
        Init(k, dimension, count_array, sum_array);
        clustering(k, rows, dimension, count_array, sum_array, DataPoints, centroids_list);
        delta = calculate_delta(k, dimension, centroids_list, sum_array);
        iter++;
    }
    output = fopen(output_filename,"w");


    output = fopen(output_filename,"w");
    for(i=0;i<k;i++){
        for ( j = 0; j < dimension; j++) {
            fprintf(output,"%.4f",centroids_list[i][j]);
            if(j<dimension-1){
                fprintf(output,",");
            }
        }
        fprintf(output,"\n");
    }

    fclose(output);
    for(i=0 ; i<k ; i++){
        free(centroids_list[i]);
        free((sum_array[i]));
    }
    free(centroids_list);
    for(i=0 ; i<rows ; i++){
        free(DataPoints[i]);
    }
    free(DataPoints);
    free(sum_array);
    free(count_array);
    return 0;

}

int main(int argc, char** argv) {
    int Valid;
    Valid = valid(argc,argv);

    if(!Valid){
        printf("Invalid Input!");
        return 1;
    }
    /*All arguments provided*/
    if (argc ==5){
        kmeans(atoi(argv[1]),atoi(argv[2]) , argv[3], argv[4]);
    }
    /*Argument "Max_iter missed*/
    if (argc==4){
        kmeans(atoi(argv[1]),200 , argv[2], argv[3]);
    }
    return 0;
}
